### PR TITLE
Enable leader election by default.

### PR DESCRIPTION
**Pull request by mattmoor**
This consolidates the core of sharedmain around the new leaderelection logic, which will now be **enabled by default**.

This can now be disabled with `--disable-ha` or by passing `sharedmain.WithHADisabled(ctx)` to `sharedmain.MainWithConfig`.

/assign @pmorie @vagababov @markusthoemmes

### CODE REVIEW

File: leaderelection/config.go [hunk 1]
```diff
@@ -84,45 +79,42 @@ func NewConfigFromConfigMap(configMap *corev1.ConfigMap) (*Config, error) {
...
-	RetryPeriod       time.Duration
+	ResourceLock  string
+	Buckets       uint32
+	LeaseDuration time.Duration
+	RenewDeadline time.Duration
+	RetryPeriod   time.Duration
+
+	// This field is deprecated and will be removed once downstream
+	// repositories have removed their validation of it.
+	// TODO(mattmoor): DO NOT SUBMIT open an issue to track.
```
Comment (mattmoor) [inline leaderelection/config.go:90]: Found "DO NOT SUBMIT".

File: injection/sharedmain/main.go [hunk 2]
```diff
@@ -118,100 +112,56 @@ func GetLeaderElectionConfig(ctx context.Context) (*kle.Config, error) {
...
+// MainWithContext runs the generic main flow for controllers and
+// webhooks. Use MainWithContext if you do not need to serve webhooks.
+func MainWithContext(ctx context.Context, component string, ctors ...injection.ControllerConstructor) {
 
-	run := func(ctx context.Context) {
-		cmw := SetupConfigMapWatchOrDie(ctx, logger)
-		controllers, _ := ControllersAndWebhooksFromCtors(ctx, cmw, ctors...)
-		WatchLoggingConfigOrDie(ctx, cmw, logger, atomicLevel, component)
-		WatchObservabilityConfigOrDie(ctx, cmw, profilingHandler, logger, component)
+	// This is a
```
Comment (vagababov) [inline injection/sharedmain/main.go:133]: unterminated comment.

File: injection/sharedmain/main.go [hunk 3]
```diff
@@ -118,100 +112,56 @@ func GetLeaderElectionConfig(ctx context.Context) (*kle.Config, error) {
...
 }
 
-// WebhookMainWithContext runs the generic main flow for controllers and
-// webhooks. Use MainWithContext if you do not need to serve webhooks.
-func WebhookMainWithContext(ctx context.Context, component string, ctors ...injection.ControllerConstructor) {
-	WebhookMainWithConfig(ctx, component, ParseAndGetConfigOrDie(), ctors...)
+// IsHADisabled checks the context for the desired to disabled leader elector.
+func IsHADisabled(ctx context.Context) bool {
+	val := ctx.Value(haDisabledKey{})
+	return val != nil
```
Comment (vagababov) [inline injection/sharedmain/main.go:159]: just 1 line?

Comment (mattmoor) [reply injection/sharedmain/main.go]: done

File: leaderelection/config.go (hunk 1 above)
Comment (mattmoor) [inline leaderelection/config.go:90]: Found "DO NOT SUBMIT".

File: injection/sharedmain/main.go (hunk 2 above)
Comment (mattmoor) [reply injection/sharedmain/main.go]: fixed

**Comment by mattmoor** [comment]
/hold

While I stage things

**Comment by mattmoor** [comment]
Downstream serving PR: https://github.com/knative/serving/pull/8602
Downstream eventing PR: https://github.com/knative/eventing/pull/3564

File: injection/sharedmain/main.go [hunk 4]
```diff
@@ -118,100 +112,55 @@ func GetLeaderElectionConfig(ctx context.Context) (*kle.Config, error) {
...
+// MainWithContext runs the generic main flow for controllers and
+// webhooks. Use MainWithContext if you do not need to serve webhooks.
+func MainWithContext(ctx context.Context, component string, ctors ...injection.ControllerConstructor) {
 
-	run := func(ctx context.Context) {
-		cmw := SetupConfigMapWatchOrDie(ctx, logger)
-		controllers, _ := ControllersAndWebhooksFromCtors(ctx, cmw, ctors...)
-		WatchLoggingConfigOrDie(ctx, cmw, logger, atomicLevel, component)
-		WatchObservabilityConfigOrDie(ctx, cmw, profilingHandler, logger, component)
+	// TODO(mattmoor): Remove this once HA is stable.
```
Comment (vagababov) [inline injection/sharedmain/main.go:133]: issue rather?

File: injection/sharedmain/main.go [hunk 5]
```diff
@@ -118,100 +112,55 @@ func GetLeaderElectionConfig(ctx context.Context) (*kle.Config, error) {
...
+	return ctx.Value(haDisabledKey{}) != nil
 }
 
-// WebhookMainWithConfig runs the generic main flow for controllers and webhooks
-// with the given config. Use MainWithConfig if you do not need to serve
-// webhooks.
-func WebhookMainWithConfig(ctx context.Context, component string, cfg *rest.Config, ctors ...injection.ControllerConstructor) {
+// MainWithConfig runs the generic main flow for controllers and webhooks
+// with the given config.
+func MainWithConfig(ctx context.Context, component string, cfg *rest.Config, ctors ...injection.ControllerConstructor) {
```
Comment (yanweiguo) [inline injection/sharedmain/main.go:163]: If leader election mode is enabled, before this change:
- webhook is running on bucket level leader election mode.
- other controllers are running on pod level leader election mode.

After this change, all controllers are running on bucket level leader election mode. Is my understading right?

Comment (mattmoor) [reply injection/sharedmain/main.go:163]: yes

**Comment by mattmoor** [comment]
I pushed one additional change that was uncovered by the chaos testing downstream on serving.  With this, I've had at least one clean run of serving (rerunning now) and just pushed it to the eventing PR.

File: injection/sharedmain/main.go [hunk 6]
```diff
@@ -118,100 +112,55 @@ func GetLeaderElectionConfig(ctx context.Context) (*kle.Config, error) {
 	return kle.NewConfigFromConfigMap(leaderElectionConfigMap)
```
Comment (vaikas) [inline injection/sharedmain/main.go:112]: nit, since there's no longer an import conflict, maybe kle -> leaderelection and drop the kle from the import?

**Comment by vaikas** [comment]
/lgtm
/approve

/hold
Just a hold if you want to address the nit, up to you and remove the hold if you want to address later/never :)

File: injection/sharedmain/main.go (hunk 6 above)
Comment (mattmoor) [reply injection/sharedmain/main.go]: Done

**Comment by mattmoor** [comment]
/hold cancel

I am going to land this now, and after lunch I'll start rebasing and landing the downstream PRs.

**Pull request merged**